// @sweet-js/helpers/maybe.js
'lang sweet.js';

export class Maybe {
  // $FlowFixMe: computed properties still not supported in flow
  static ['fantasy-land/of'](v) {
    return Maybe.of(v);
  }

  static of(v) {
    return new Just(v);
  }

  // $FlowFixMe: computed properties still not supported in flow
  static ['fantasy-land/empty']() {
    return Maybe.empty();
  }

  static empty() {
    return new Nothing();
  }

  // $FlowFixMe: computed properties still not supported in flow
  static ['fantasy-land/zero']() {
    return Maybe.zero();
  }

  static zero() {
    return new Nothing();
  }

  static isJust(v) {
    return v instanceof Just;
  }

  static isNothing(v) {
    return v instanceof Nothing;
  }

  // $FlowFixMe: computed properties still not supported in flow
  ['fantasy-land/map'](f) {
    return this.map(f);
  }

  map(f) {
    throw new Error();
  }

  // $FlowFixMe: computed properties still not supported in flow
  ['fantasy-land/ap'](other) {
    return this.ap(other);
  }

  ap(other) {
    throw new Error();
  }

  // $FlowFixMe: computed properties still not supported in flow
  ['fantasy-land/chain'](f) {
    return this.chain(f);
  }

  chain(f) {
    throw new Error();
  }
}

export class Nothing extends Maybe {
  map(f) {
    return Maybe.empty();
  }

  ap(other) {
    return Maybe.empty();
  }

  chain(f) {
    return Maybe.empty();
  }
}

export class Just extends Maybe {

  constructor(v) {
    super();
    this.value = v;
  }

  map(f) {
    return new Just(f(this.value));
  }

  ap(other) {
    if (other instanceof Just) {
      return Maybe.of(other.value(this.value));
    }
    return Maybe.empty();
  }

  chain(f) {
    return f(this.value);
  }
}
